### PR TITLE
Refactoring forms

### PR DESCRIPTION
**Pull request by marcospereira**
## Purpose

This refactors `Form.bind` method into a set of smaller and more digestible methods.

## References

Follow up of the work done by @mkurz in #7038. See discussion there about this refactoring.

### CODE REVIEW

**Comment by benmccann** [comment]
Just a heads up that the build is failing right now. I've submitted a PR to fix

**Comment by marcospereira** [comment]
@benmccann probably not related to the changes here. Triggering the build again made it green.

File: framework/src/play-java-forms/src/main/java/play/data/Form.java [hunk 1]
```diff
@@ -377,135 +362,180 @@ protected String getMessageForConstraintViolation(ConstraintViolation<Object> vi
...
+                if (v.getParameterCount() == 0) {
+                    globalError = v.invoke(result.getTarget());
+                    Logger.warn("The \"validate\" method in class \"{}\" is deprecated since Play 2.6. " +
+                                    "To migrate to class-level constraints see https://www.playframework.com/documentation/2.6.x/Migration26#java-form-changes " +
+                                    "and https://www.playframework.com/documentation/2.6.x/JavaForms#Advanced-validation",
+                            result.getTarget().getClass().getName());
+                }
+            } catch (NoSuchMethodException e) {
+                // do nothing
+            } catch (Throwable e) {
```
Comment (gmethvin) [inline framework/src/play-java-forms/src/main/java/play/data/Form.java]: Which checked exceptions are we rethrowing? Might be better to only rethrow those specific exception types as   it documents which ones you're ignoring.

File: framework/src/play-java-forms/src/main/java/play/data/Form.java [hunk 2]
```diff
@@ -347,28 +348,12 @@ protected String getMessageForConstraintViolation(ConstraintViolation<Object> vi
...
+    private DataBinder dataBinder(String... allowedFields) {
         DataBinder dataBinder;
-        Map<String, String> objectData = data;
         if (rootName == null) {
             dataBinder = new DataBinder(blankInstance());
         } else {
             dataBinder = new DataBinder(blankInstance(), rootName);
-            objectData = new HashMap<>();
-            for (String key: data.keySet()) {
-                if (key.startsWith(rootName + ".")) {
```
Comment (mkurz) [inline framework/src/play-java-forms/src/main/java/play/data/Form.java]: This line is missing now. Is it really safe to skip it?

Comment (marcospereira) [inline framework/src/play-java-forms/src/main/java/play/data/Form.java]: Probably it is not safe to skip it. If it is not failing, then we are missing a test here.

**Comment by marcospereira** [comment]
@gmethvin and @mkurz both points addressed.

File: framework/src/play-java-forms/src/main/java/play/data/Form.java (hunk 2 above)
Comment (marcospereira) [inline framework/src/play-java-forms/src/main/java/play/data/Form.java]: It is back here now:

https://github.com/playframework/playframework/pull/7132/files#diff-aa7d1890e0e0a4bdad1abbc67b2ecfc9R372

With additional tests.

**Pull request merged**